hathor/websocket/streamer.py

```diff
@@ -0,0 +1,221 @@
+# Copyright 2024 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import deque
+from collections.abc import AsyncIterable
+from typing import TYPE_CHECKING, Any, AsyncIterator
+
+from twisted.internet.defer import Deferred
+from twisted.internet.interfaces import IPushProducer
+from twisted.internet.task import deferLater
+from zope.interface import implementer
+
+from hathor.manager import HathorManager
+from hathor.types import AddressB58
+from hathor.util import json_dumpb
+
+if TYPE_CHECKING:
+    from hathor.websocket.protocol import HathorAdminWebsocketProtocol
+
+
+class ManualAddressSequencer(AsyncIterable[tuple[int, AddressB58]]):
+    def __init__(self) -> None:
+        self.pending_addresses: deque[AddressB58] = deque()
+        self.stop = False
+        self.await_items: Deferred | None = None
+
+    def __aiter__(self) -> AsyncIterator[tuple[int, AddressB58]]:
+        return self._async_iter()
+
+    def _resume_iter(self) -> None:
+        if self.await_items is None:
+            return
+        self.await_items.callback(None)
+
+    def add_addresses(self, addresses: list[AddressB58], last: bool) -> None:
+        self.pending_addresses.extend(addresses)
+        if last:
+            self.stop = True
+        self._resume_iter()
+
+    async def _async_iter(self) -> AsyncIterator[tuple[int, AddressB58]]:
+        idx = 0
```

Comment:
This assumes that the index starts at 0 and the addresses are all in order, while this is the expected behavior maybe the wallet should send the index with each address.
So that the fullnode does not have to guess the address index.

Reply:
Fixed!